### PR TITLE
Lift artificial restrictions on `ConstantAnnotation`s

### LINKED ISSUE #11724: ConstantAnnotation has unnecessary restrictions due to its Java heritage
For example, named arguments need to be used unless therese one single parameter called `value`. We should implement this restirction only for Java annotations.

```
scala> class a(x: Int) extends annotation.ConstantAnnotation
defined class a

scala> @a(1) val x = 0
          ^
       error: arguments to Java annotations or subclasses of ConstantAnnotation have to be supplied as named arguments
        ^
       error: annotation a is missing argument x

scala> class a(value: Int) extends annotation.ConstantAnnotation
defined class a

scala> @a(1) val x = 0
x: Int = 0
```

### PR DESCRIPTION
**Pull request by lrytz**
This PR drops some artificial restrictions for annotations extending `ConstantAnnotation`
  - `@constAnn(arg1, arg2)` is now allowed without using named arguments
  - `@constAnn(arg)` is now allowed for a generic annotation `class constAnn[T](x: T) extends ConstantAnnotation`

This accomplished by typechecking `ConstantAnnotation`s normally, not like Java annotations.

So far we handled ConstantAnnotations the same as Java annotations.
Instead of typechecking the full parse tree `Apply(New(annot), args)`,
the `New` and the individual `args` were typed individually. This is
needed for Java annotations as there's no corresponding constructor.
But for ConstantAnnotations we can just type check them normally and
then extract the constants.

Fixes https://github.com/scala/bug/issues/11724.

Context: https://github.com/scala/scala/pull/9336#discussion_r532669420

### CODE REVIEW

File: test/files/neg/nested-annotation.check [hunk 1]
```diff
@@ -1,4 +1,7 @@
 nested-annotation.scala:8: error: nested classfile annotations must be defined in java; found: inline
   @ComplexAnnotation(new inline) def bippy(): Int = 1
                      ^
-1 error
+nested-annotation.scala:11: error: no arguments allowed for nullary constructor SuppressWarnings: (): SuppressWarnings
+  @ComplexAnnotation(new SuppressWarnings(Array("blup"))) def huppy(): Int = 2
+                                               ^
```
Comment (lrytz) [inline test/files/neg/nested-annotation.check]: 🤔

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 2]
```diff
@@ -4026,13 +4033,10 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
         } else {
           // TODO: annType may have undetermined type params for Scala ConstantAnnotations, see scala/bug#11724.
```
Comment (dwijnand) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: stale

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 3]
```diff
@@ -4124,12 +4128,38 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
...
+              case (arg, param) =>
+                val origArg = arg match {
+                  case Ident(n) => namedArgs.getOrElse(n, arg)
+                  case _ => arg
+                }
+                (param.name, tree2ConstArg(origArg, param.tpe.resultType))
+            }
+            if (hasError) ErroneousAnnotation
+            else if (unmappable) UnmappableAnnotation
+                 else AnnotationInfo(info.atp, Nil, assocs.collect({ case (n, Some(arg)) => (n, arg) })).setOriginal(info.original).setPos(info.pos)
```
Comment (dwijnand) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: (extra indent)

**Review by dwijnand — APPROVED** [review]
Shame this has to undo what typer did to named/default args, making it net more LOC in the compiler.  But I guess the effect is progress.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 3 above)
Comment (lrytz) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala]: IntelliJ bug...

**Comment by dwijnand** [comment]
It's not big, but it allows for more things, so I feel we should bullet point this for users, in the release notes.

**Pull request merged**